### ISSUE TITLE
Different idmaps per container with security.idmap.isolated

### ISSUE THREAD

**Issue by vwvol** [lxc/lxd#3208]
# Required information

 * Distribution: Ubuntu
 * Distribution version: 16.10
 * The output of "lxc info" or if that fails:
   * Kernel version: 4.8.0-46-generic
   * LXC version: 2.0.7
   * LXD version: 2.12
   * Storage backend in use: dir

# Issue description

An error occurs while trying to configure different idmaps per container with security.idmap.isolated using subuid/subgid range equal 65536.
But it works when subuid/subgid range is equal 1000000. I'm not sure if this is a bug but using such large range value looks more like a workaround or not documented feature.

# Steps to reproduce

 1. Install lxd and lxd-client packages from yakkety-backports.
 2. Create storage and try to start container with security.idmap.isolated=true using default subuid/subgid files. Get an error.

```
root@ubuntu:~# cat /etc/subuid
user:100000:65536
lxd:165536:65536
root:165536:65536
root@ubuntu:~# cat /etc/subgid
user:100000:65536
lxd:165536:65536
root:165536:65536
root@ubuntu:~# lxc storage create default dir
Storage pool default created
root@ubuntu:~# lxc profile device add default root disk path=/ pool=default
Device root added to default
root@ubuntu:~# lxc init images:debian/jessie deb01 -c security.idmap.isolated=true
Creating deb01
error: Not enough uid/gid available for the container.
```

 3. Edit both /etc/subuid and /etc/subgid files, restart lxd and try again. Everything works fine.

```
root@ubuntu:~# cat /etc/subuid
user:100000:65536
lxd:165536:65536
root:165536:1000000
root@ubuntu:~# cat /etc/subgid
user:100000:65536
lxd:165536:65536
root:165536:1000000
root@ubuntu:~# systemctl restart lxd.service
root@ubuntu:~# lxc init images:debian/jessie deb01 -c security.idmap.isolated=true
Creating deb01

The container you are starting doesn't have any network attached to it.
  To create a new network, use: lxc network create
  To attach a network to a container, use: lxc network attach

root@ubuntu:~# lxc start deb01
root@ubuntu:~#
```

**Comment by stgraber** [comment]
That's expected behavior.

security.idmap.isolated means that LXD will allocate 65536 uid and gid PER CONTAINER.
Your configuration shows that you had 65536 uid and gid TOTAL for LXD to use.

So LXD obviously couldn't find enough uid and gid for your newly created container and failed with the error telling you as much.

Once you increased the number of available uid and gid from 65536 to 1000000, LXD now had enough uid/gid to use for container creation (you'll run out again after 15 containers or so).

**Comment by stgraber** [comment]
Ah, one thing worth clarifying. security.idmap.isolated causes LXD to use a map other than the first, so even though you had 65536 configured, LXD wouldn't use those for a security.idmap.isolated containers, it'd use the first chunk only for containers which don't have that key set.

So if you want enough uid/gid to run ONE container using security.idmap.isolated and any number of containers without it set, you'd need a map of 131072 uid/gid.
This isn't a very likely scenario though as people who use security.idmap.isolated usually want it for all their containers at which point you need 65536 uid/gid a piece and so will bump your allocation to a very large value to give LXD some breathing room.
